### ISSUE TITLE
Different result when using self-defined objective

### ISSUE THREAD

**Issue by xdaniel9** [microsoft/LightGBM#4077]
sample code:
```
lgb_params = {
            "verbose": 1,
            "objective": "regression",
            "metric": "None",

            "boosting": "gbdt",
            "learning_rate": 0.01,
            "num_leaves": 8,
            "max_depth": 4,
            "min_data_in_leaf": 64,
            "max_bin": 128,

            "bagging_fraction": 1.0,
            "feature_fraction": 1.0,
            
            "lambda_l1": 0,
            "lambda_l2": 0,
            
            "seed": 0,
            "first_metric_only": True
        }
X = np.random.randn(2000, 300)
y = np.random.randn(2000,) + (np.random.randn(300)*X).sum(axis=1) + 0

def fobj(pred, data):
    return (pred - data.label), np.ones(len(pred))

# self defined
dataset = lgb.Dataset(X, y)
model1 = lgb.train(
        lgb_params, dataset,
        fobj=fobj,
        num_boost_round=1,
        init_model=None,
    )

# default
dataset = lgb.Dataset(X, y)
model2 = lgb.train(
        lgb_params, dataset,
        fobj=None,
        num_boost_round=1,
        init_model=None,
    )

model1.predict(X)
# array([ 0.02129382,  0.02129382,  0.08837065, ..., -0.03432668, 0.08837065, -0.01270184])

model2.predict(X)
# array([-0.22792455, -0.22792455, -0.16084771, ..., -0.28354505,-0.16084771, -0.2619202 ])
```

Since the data and seed is the same, I except two models gives the same result

**Comment by shiyu1994** [comment]
@xdaniel9 Thanks for using LightGBM. The customized objective function is the same as `l2` loss. But when using the internal `regression` objective (`l2` loss), LightGBM will boost from the average label, instead of 0. That means, the initial predict score of all samples will be set to the average label. But with customized function, since LightGBM don't know how to set a proper initial predict score for that objective, the default initial score will be 0. 

That causes the difference of the two results.

The same problem has been mentioned in #2558,

**Comment by StrikerRUS** [comment]
Although according to the code this issue is indeed about `boost_from_average` inconsistency, I believe it is worth to mention one very similar problem where results for self-defined objective are different compared to the built-in (**not only**) `mae` objective.
It might be useful later for those who will come here from search engines.
https://github.com/microsoft/LightGBM/issues/3584#issuecomment-731598701

@xdaniel9 
Adding 
```
    "boost_from_average": False
```
to `lgb_params` makes results identical.
